rustls/src/conn/unbuffered.rs

```diff
@@ -316,50 +316,51 @@ impl<Data> fmt::Debug for ConnectionState<'_, '_, Data> {
 
 /// Application data is available
 pub struct ReadTraffic<'c, 'i, Data> {
-    _conn: &'c mut UnbufferedConnectionCommon<Data>,
+    conn: &'c mut UnbufferedConnectionCommon<Data>,
     // for forwards compatibility; to support in-place decryption in the future
     _incoming_tls: &'i mut [u8],
-    chunk: Vec<u8>,
-    taken: bool,
+
+    // owner of the latest chunk obtained in `next_record`, as borrowed by
+    // `AppDataRecord`
+    chunk: Option<Vec<u8>>,
 }
 
 impl<'c, 'i, Data> ReadTraffic<'c, 'i, Data> {
-    fn new(
-        _conn: &'c mut UnbufferedConnectionCommon<Data>,
-        _incoming_tls: &'i mut [u8],
-        chunk: Vec<u8>,
-    ) -> Self {
+    fn new(conn: &'c mut UnbufferedConnectionCommon<Data>, _incoming_tls: &'i mut [u8]) -> Self {
         Self {
-            _conn,
+            conn,
             _incoming_tls,
-            chunk,
-            taken: false,
+            chunk: None,
         }
     }
 
     /// Decrypts and returns the next available app-data record
     // TODO deprecate in favor of `Iterator` implementation, which requires in-place decryption
     pub fn next_record(&mut self) -> Option<Result<AppDataRecord<'_>, Error>> {
-        if self.taken {
-            None
-        } else {
-            self.taken = true;
-            Some(Ok(AppDataRecord {
+        self.chunk = self
```

Comment:
On `ReadTraffic` `Drop`, should we put back the `chunk` if `Some`?

(I guess this is covered by your commit message saying that `next_record()` remains one-shot.)

Reply:
No. This is the point at which the chunk is consumed. We need to save it in `self.chunk` to get a place for `AppDataRecord` to borrow from.